### PR TITLE
Move some new state tests to pytest

### PR DESCRIPTION
**Pull request by krionbsd**
### What does this PR do?
Move some new state tests to pytest

### CODE REVIEW

File: tests/pytests/unit/states/test_git.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def test_latest_no_diff_for_bare_repo(tmp_path):
+    """
+    This test ensures that we don't attempt to diff when cloning a repo
+    using either bare=True or mirror=True.
+    """
+    name = "https://foo.com/bar/baz.git"
+    gitdir = os.path.join(tmp_path, "refs")
+    isdir_mock = MagicMock(side_effect=lambda path: DEFAULT if path != gitdir else True)
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_git.py:26]:
```suggestion
    gitdir = tmp_path / "refs"
    isdir_mock = MagicMock(side_effect=lambda path: DEFAULT if path != str(gitdir) else True)
```

File: tests/pytests/unit/states/test_git.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        "git.rev_parse": MagicMock(side_effect=git_state.CommandExecutionError()),
+        "git.revision": MagicMock(return_value=local_head),
+        "git.version": MagicMock(return_value="1.8.3.1"),
+    }
+    with patch("os.path.isdir", isdir_mock), patch.dict(
+        git_state.__salt__, dunder_salt
+    ):
+        result = git_state.latest(
+            name=name, target=tmp_path, mirror=True,  # mirror=True implies bare=True
+        )
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_git.py:55]:
```suggestion
        result = git_state.latest(
            name=name, target=str(tmp_path), mirror=True,  # mirror=True implies bare=True
        )
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
Changes required for Py3.5 compat

File: tests/pytests/unit/states/test_git.py (hunk 1 above)
Comment (krionbsd) [reply tests/pytests/unit/states/test_git.py]: Fixed

File: tests/pytests/unit/states/test_git.py (hunk 2 above)
Comment (krionbsd) [reply tests/pytests/unit/states/test_git.py]: Fixed

File: tests/pytests/unit/states/test_esxdatacenter.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            esxdatacenter.__salt__,
+            {
+                "vsphere.get_proxy_type": MagicMock(),
+                "vsphere.get_service_instance_via_proxy": MagicMock(
+                    return_value=mock_si
+                ),
+                "vsphere.list_datacenters_via_proxy": MagicMock(return_value=[mock_dc]),
+                "vsphere.disconnect": MagicMock(),
+            },
+        )
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_esxdatacenter.py:50]:
```suggestion
@pytest.fixture(scope="function")
def policy_set():
    patcher_1 = patcher_2 = None
    try:
        patcher_1 = patch.dict(esxdatacenter.__opts__, {"test": False})
        patcher_1.start()

        mock_si = MagicMock()
        mock_dc = MagicMock()

        patcher_2 = patch.dict(
            esxdatacenter.__salt__,
            {
                "vsphere.get_proxy_type": MagicMock(),
                "vsphere.get_service_instance_via_proxy": MagicMock(
                    return_value=mock_si
                ),
                "vsphere.list_datacenters_via_proxy": MagicMock(return_value=[mock_dc]),
                "vsphere.disconnect": MagicMock(),
            },
        )
        patcher_2.start()
        yield
    finally:
        if patcher_1:
            patcher_1.stop()
        if patcher_2:
            patcher_2.stop()
```
This way it should still work, can you please try?

Comment (krionbsd) [reply tests/pytests/unit/states/test_esxdatacenter.py]: Yep, it works, applied these changes

File: tests/pytests/unit/states/test_kernelpkg.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                assert kernelpkg.__salt__["kernelpkg.upgrade"]
+
+            with patch.dict(kernelpkg.__opts__, {"test": True}):
+                kernelpkg.__salt__["kernelpkg.upgrade"].reset_mock()
+                ret = kernelpkg.latest_installed(name=STATE_NAME)
+                assert ret["name"] == STATE_NAME
+                assert ret["result"] is None
+                assert ret["changes"] == {}
+                assert isinstance(ret["comment"], str)
+                kernelpkg.__salt__["kernelpkg.upgrade"].assert_not_called()
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_kernelpkg.py:51]:
```suggestion
    with patch.dict(kernelpkg.__salt__, {"kernelpkg.list_installed": installed, "kernelpkg.latest_available": upgrade}):
        with patch.dict(kernelpkg.__opts__, {"test": False}):
            kernelpkg.__salt__["kernelpkg.upgrade"].reset_mock()
            ret = kernelpkg.latest_installed(name=STATE_NAME)
            assert ret["name"] == STATE_NAME
            assert ret["result"]
            assert isinstance(ret["changes"], dict)
            assert isinstance(ret["comment"], str)
            assert kernelpkg.__salt__["kernelpkg.upgrade"]

        with patch.dict(kernelpkg.__opts__, {"test": True}):
            kernelpkg.__salt__["kernelpkg.upgrade"].reset_mock()
            ret = kernelpkg.latest_installed(name=STATE_NAME)
            assert ret["name"] == STATE_NAME
            assert ret["result"] is None
            assert ret["changes"] == {}
            assert isinstance(ret["comment"], str)
            kernelpkg.__salt__["kernelpkg.upgrade"].assert_not_called()
```

File: tests/pytests/unit/states/test_kernelpkg.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+
+def test_latest_installed_at_latest():
+    """
+    Test - latest_installed when no upgrade is available
+    """
+    installed = MagicMock(return_value=KERNEL_LIST)
+    upgrade = MagicMock(return_value=KERNEL_LIST[-1])
+    with patch.dict(kernelpkg.__salt__, {"kernelpkg.list_installed": installed}):
+        with patch.dict(kernelpkg.__salt__, {"kernelpkg.latest_available": upgrade}):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_kernelpkg.py:61]: Same as above, can be joined on a single patch call

File: tests/pytests/unit/states/test_keystore.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+    # with patch.dict(keystore.__opts__, {'test': False}):
+    with patch.dict(
+        keystore.__salt__, {"keystore.list": MagicMock(return_value=cert_return)}
+    ):
+        with patch.dict(
+            keystore.__salt__,
+            {"x509.read_certificate": MagicMock(return_value=x509_return)},
+        ):
+            assert keystore.managed(name, passphrase, entries) == state_return
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:95]:
```suggestion
    with patch.dict(
        keystore.__salt__, {"keystore.list": MagicMock(return_value=cert_return), "x509.read_certificate": MagicMock(return_value=x509_return)}
    ):
        assert keystore.managed(name, passphrase, entries) == state_return
```

File: tests/pytests/unit/states/test_keystore.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+    with patch.dict(keystore.__opts__, {"test": True}):
+        with patch.dict(
+            keystore.__salt__, {"keystore.list": MagicMock(return_value=cert_return)},
+        ):
+            with patch.dict(
+                keystore.__salt__,
+                {"x509.read_certificate": MagicMock(return_value=x509_return)},
+            ):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:104]: Same as above, single patch

File: tests/pytests/unit/states/test_keystore.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+    with patch.dict(keystore.__opts__, {"test": True}):
+        with patch.dict(
+            keystore.__salt__, {"keystore.list": MagicMock(return_value=cert_return)},
+        ):
+            with patch.dict(
+                keystore.__salt__,
+                {"x509.read_certificate": MagicMock(return_value=x509_return)},
+            ):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:194]: Single patch

File: tests/pytests/unit/states/test_keystore.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+        with patch.dict(
+            keystore.__salt__,
+            {"x509.read_certificate": MagicMock(return_value=x509_return)},
+        ):
+            with patch.dict(
+                keystore.__salt__, {"keystore.remove": MagicMock(return_value=True)}
+            ):
+                with patch.dict(
+                    keystore.__salt__, {"keystore.add": MagicMock(return_value=True)},
+                ):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:209]: Single patch

File: tests/pytests/unit/states/test_keystore.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+
+    with patch.dict(keystore.__opts__, {"test": True}):
+        assert keystore.managed(name, passphrase, entries) == test_return
+
+    with patch.dict(
+        keystore.__salt__, {"keystore.remove": MagicMock(return_value=True)}
+    ):
+        with patch.dict(
+            keystore.__salt__, {"keystore.add": MagicMock(return_value=True)}
+        ):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:260]: Single patch

File: tests/pytests/unit/states/test_keystore.py [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+    with patch.dict(keystore.__opts__, {"test": True}):
+        with patch.dict(
+            keystore.__salt__, {"keystore.list": MagicMock(return_value=cert_return)},
+        ):
+            with patch.dict(
+                keystore.__salt__,
+                {"x509.read_certificate": MagicMock(return_value=x509_return)},
+            ):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:350]: single patch

File: tests/pytests/unit/states/test_keystore.py [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+        with patch.dict(
+            keystore.__salt__,
+            {"x509.read_certificate": MagicMock(return_value=x509_return)},
+        ):
+            with patch.dict(
+                keystore.__salt__, {"keystore.remove": MagicMock(return_value=True)}
+            ):
+                with patch.dict(
+                    keystore.__salt__, {"keystore.add": MagicMock(return_value=True)},
+                ):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:368]: Single patch

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/states/test_glusterfs.py [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+
+            comt = "Volume {} already exists".format(name)
+            ret.update({"comment": comt, "changes": {}})
+            assert glusterfs.volume_present(name, bricks, start=False) == ret
+        with patch.dict(glusterfs.__opts__, {"test": True}):
+            comt = "Volume {} already exists".format(name)
+            ret.update({"comment": comt, "result": None})
+            assert glusterfs.volume_present(name, bricks, start=False) == ret
+
+            comt = ("Volume {0} already exists" + " and will be started").format(name)
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_glusterfs.py:131]:
```suggestion
            comt = "Volume {} already exists and will be started".format(name)
```

File: tests/pytests/unit/states/test_glusterfs.py [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+            ret.update({"comment": comt, "result": None})
+            assert glusterfs.volume_present(name, bricks, start=True) == ret
+
+            mock_list.return_value = []
+            comt = "Volume {} will be created".format(name)
+            ret.update({"comment": comt, "result": None})
+            assert glusterfs.volume_present(name, bricks, start=False) == ret
+
+            comt = ("Volume {0} will be created" + " and started").format(name)
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_glusterfs.py:140]:
```suggestion
            comt = "Volume {} will be created and started".format(name)
```

File: tests/pytests/unit/states/test_keystore.py (hunk 7 above)
Comment (s0undt3ch) [reply tests/pytests/unit/states/test_keystore.py:104]: Still applies, single patch

File: tests/pytests/unit/states/test_keystore.py [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+            "x509.read_certificate": MagicMock(return_value=x509_return),
+        },
+    ):
+        with patch.dict(
+            keystore.__salt__,
+            {
+                "keystore.remove": MagicMock(return_value=True),
+                "keystore.add": MagicMock(return_value=True),
+            },
+        ):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:209]: Single patch

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/states/test_kapacitor.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+        "/tmp/script.tick",
+        database="testdb",
+        retention_policy="default",
+        task_type="stream",
+        dbrps=["testdb2.default_rp", "testdb.default"],
+    )
+    enable_mock.assert_called_once_with("testname")
+    assert "TICKscript diff" in ret["changes"]
+    assert "enabled" in ret["changes"]
+    assert True is ret["changes"]["enabled"]["new"]
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_kapacitor.py:91]:
```suggestion
    assert ret["changes"]["enabled"]["new"] is True
```

Comment (s0undt3ch) [reply tests/pytests/unit/states/test_kapacitor.py:91]: and all other similar occurrences in this file

File: tests/pytests/unit/states/test_keystore.py [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+    ]
+
+    state_return = {
+        "name": name,
+        "changes": {},
+        "result": True,
+        "comment": "No changes made.\n",
+    }
+
+    # with patch.dict(keystore.__opts__, {'test': False}):
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:87]: Comment no longer required right?

File: tests/pytests/unit/states/test_keystore.py [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+            "keystore.list": MagicMock(return_value=cert_return),
+            "x509.read_certificate": MagicMock(return_value=x509_return),
+            "keystore.remove": MagicMock(return_value=True),
+            "keystore.add": MagicMock(return_value=True),
+        },
+    ):
+        assert keystore.managed(name, passphrase, entries) == state_return
+
+
+@patch("os.path.exists", MagicMock(return_value=False))
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/test_keystore.py:208]: Let's not use the `@patch` decorator please. Use `with patch()` inside the test function or use a fixture to do it outside of the test function.
https://engblog.nextdoor.com/what-your-mocks-do-when-you-aren-t-looking-b278e0d9e201

Comment (s0undt3ch) [reply tests/pytests/unit/states/test_keystore.py:208]: The same for all other patch decorators in this module

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Pull request closed without merging**